frontend/test/metabase/parameters/components/Parameters/syncQueryParamsWithURL.unit.spec.js

```diff
@@ -2,9 +2,215 @@ import Field from "metabase-lib/lib/metadata/Field";
+          syncQueryParamsWithURL(props);
+          expect(setParameterValue).toHaveBeenCalledWith("id1", "2021");
+          expect(setParameterValue).toHaveBeenCalledWith("id2", "CA");
+        });
+      });
+    });
+
+    describe("for public questions", () => {
+      describe("when parameters length is 0", () => {
+        const props = {
+          parameters: [],
+          query: {
+            createdAt: "2021",
+          },
+          setParameterValue,
+        };
+
+        it("calls setParameterValue with empty object as argument", () => {
+          syncQueryParamsWithURL(props);
+          expect(setParameterValue).toHaveBeenCalledTimes(1);
+          expect(setParameterValue).toHaveBeenCalledWith({});
+        });
+      });
+
+      describe("when query has no key that matches a parameter slug", () => {
+        const props = {
+          parameters: [
+            {
+              id: "id",
+              slug: "slugNotKeyInQuery",
+            },
+          ],
+          query: {
+            createdAt: "2021",
+          },
+          setParameterValue,
+        };
+
+        it("calls setParameterValue with empty object as argument", () => {
+          syncQueryParamsWithURL(props);
+          expect(setParameterValue).toHaveBeenCalledWith({});
+        });
+      });
+
+      describe("when parameters length is 1", () => {
+        const props = {
+          parameters: [
+            {
+              id: "id",
+              slug: "createdAt",
+            },
+          ],
+          query: {
+            createdAt: "2021",
+          },
+          setParameterValue,
+        };
+
+        it("calls setParameterValue once", () => {
+          syncQueryParamsWithURL(props);
+          expect(setParameterValue).toHaveBeenCalledTimes(1);
+        });
+
+        it("calls setParameterValue with an object of key parameter.id and value of a parsed query param ", () => {
+          syncQueryParamsWithURL(props);
+          expect(setParameterValue).toHaveBeenCalledWith({ id: "2021" });
+        });
+      });
+
+      describe("when parameters length is 2", () => {
+        const props = {
+          parameters: [
+            {
+              id: "id1",
+              slug: "createdAt",
+            },
+            {
+              id: "id2",
+              slug: "state",
+            },
+          ],
+
+          query: {
+            createdAt: "2021",
+            state: "CA",
+          },
+          setParameterValue,
+        };
+
+        it("calls setParameterValue once", () => {
+          syncQueryParamsWithURL(props);
+          expect(setParameterValue).toHaveBeenCalledTimes(1);
+        });
+
+        it("calls setParameterValue with one object as argument, keys of parameter id and parsed param values", () => {
+          syncQueryParamsWithURL(props);
+          expect(setParameterValue).toHaveBeenCalledWith({
+            id1: "2021",
+            id2: "CA",
+          });
+        });
```

Comment:
I guess we can improve the test names here. Right now they are mostly identical to the assert statements, so they don't bring new context about the behavior we test, and a developer needs to spend some extra effort to figure out what's happening here.

I believe test names should explain _why_ the behavior should be like that, and _how_ usually visible from the code itself. Also, unit tests become much more valuable in this case as they also become some sort of a spec or doc for a function.

E.g. it can be something like that (I'm sure the wording can be better, but I hope it still represents the idea):

```js
it("doesn't change anything if there are no matching parameters", () => {
  const props = {
    parameters: [
      {
        id: "some-param",
        slug: "noMatchingQueryParam",
      },
    ],
    query: {
      someQueryParam: "noMatchingParameter",
    },
  };

  syncQueryParamsWithURL(props);

  expect(setParameterValue).not.toHaveBeenCalled();
});

it("syncs parameter with query params by slugs", () => {
  const props = {
    parameters: [
      {
        id: "created-at-param-id",
        slug: "createdAt",
      },
    ],
    query: {
      createdAt: "2021",
    },
  };

  syncQueryParamsWithURL(props);

  expect(setParameterValue).toHaveBeenCalledTimes(1);
  expect(setParameterValue).toHaveBeenCalledWith({
    "created-at-param-id": "2021",
  });
});

it("syncs multiple parameters with query params by slugs", () => {
  const props = {
    parameters: [
      {
        id: "created-at-param-id",
        slug: "createdAt",
      },
      {
        id: "state-param-id",
        slug: "state",
      },
    ],
    query: {
      createdAt: "2021",
      state: "CA",
    },
  };

  syncQueryParamsWithURL(props);

  expect(setParameterValue).toHaveBeenCalledTimes(1);
  expect(setParameterValue).toHaveBeenCalledWith({
    "created-at-param-id": "2021",
    "state-param-id": "CA",
  });
});
```

Reply:
Yup, good call, will update 🙇

Reply:
@kulyk, thank you for your review.

I've just updated the PR based on your feedback.